File: solutions/elixir/grains/lib/grains.ex
defmodule Grains do
  @doc """
  Calculate two to the power of the input minus one.
  """
  @spec square(pos_integer()) :: {:ok, pos_integer()} | {:error, String.t()}
  def square(number) when number in 1..64 do
    {:ok, 2 ** (number - 1)}
  end

  def square(_), do: {:error, "The requested square must be between 1 and 64 (inclusive)"}

  @doc """
  Adds square of each number from 1 to 64.
  """
  @spec total :: {:ok, pos_integer()}
  def total do
    Enum.reduce(1..64, 0, fn n, sum ->
      with {:ok, grains} <- square(n) do
        grains + sum
      end
    end)
    |> then(&{:ok, &1})
  end
end

File: solutions/prolog/grains/grains.pl
square(1, 1):- true, !.
square(N, Grains) :- 
    between(1, 64, N),
    Previous is N - 1,
    square(Previous, PreviousGrains),
    Grains is PreviousGrains * 2.

total(Value) :-
    numlist(1, 64, Squares),
    maplist(square, Squares, Grains),
    sum_list(Grains, Value). 